tests/test_kernel.py

```diff
@@ -597,6 +598,47 @@ def test_control_thread_priority():
     assert control_dates[-1] <= shell_dates[0]
 
 
+def test_sequential_control_messages():
+    with new_kernel() as kc:
+        msg_id = kc.execute("import time")
+        get_reply(kc, msg_id)
+
+        # Send multiple messages on the control channel.
+        # Using execute messages to vary duration.
+        sleeps = [0.6, 0.3, 0.1]
+
+        # Prepare messages
+        msgs = [
+            kc.session.msg("execute_request", {"code": f"time.sleep({sleep})"}) for sleep in sleeps
+        ]
+        msg_ids = [msg["header"]["msg_id"] for msg in msgs]
+
+        # Submit messages
+        for msg in msgs:
+            kc.control_channel.send(msg)
+
+        # Get replies
+        replies = [get_reply(kc, msg_id, channel="control") for msg_id in msg_ids]
+
+        # Check messages are processed in order, one at a time, and of a sensible duration.
+        previous_end = None
+        for reply, sleep in zip(replies, sleeps):
+            start_str = reply["metadata"]["started"]
+            if sys.version_info[:2] < (3, 11) and start_str.endswith("Z"):
+                # Python < 3.11 doesn't support "Z" suffix in datetime.fromisoformat,
+                # so use alternative timezone format.
+                # https://github.com/python/cpython/issues/80010
+                start_str = start_str[:-1] + "+00:00"
+            start = datetime.fromisoformat(start_str)
+            end = reply["header"]["date"]  # Already a datetime
+
+            if previous_end is not None:
+                assert start > previous_end
+            previous_end = end
+
+            assert end >= start + timedelta(seconds=sleep)
```

Comment:
This fails downstream in IPython with:

TypeError: '>=' not supported between instances of 'str' and 'datetime.datetime'

Reply:
These are the same symptoms that I observed at https://github.com/ipython/ipykernel/pull/1210#discussion_r1487916907. `ipyparallel` unfortunately monkey-patches jupyter_clients conversion of strings to datetimes so it is a no-op, and some versions of `pytest-asyncio` don't fully isolate the tests so sometimes some of these start and end dates come through as strings.

In this PR updating `pytest-asyncio` to 0.23.5 worked, but this is evidently not sufficient as the IPython failure already uses `pytest-asyncio-0.23.5`.

I thing the sensible approach here is to assume that in this test sometimes, outside of our direct control, the dates are strings and convert them to datetimes. Otherwise this test is not testing what it is supposed to, it is testing the types of some fields returned in messages. I'll submit a PR here ASAP.